### PR TITLE
[REF] currency_rate: Using ormcache for conversion_rate to increase performance

### PR DESCRIPTION
**Pull request by moylop260**
# Description of the issue/feature this PR addresses:
- Create 10 favorite journals
- Create 2k invoices for each one using foreign currency
   - I mean, `invoice.currency_id != invoice.company.currency_id`
- Open Menu -> Account (Opening acccount dashboard)

# Current behavior before PR:
- Too slow following `_get_rates` method:
  - https://github.com/odoo/odoo/blob/9b6073c3ea6264f59fb11e6f797f4fff200c3d5d/odoo/addons/base/models/res_currency.py#L48
    - https://github.com/odoo/odoo/blob/9b6073c3ea6264f59fb11e6f797f4fff200c3d5d/addons/account/models/account_journal_dashboard.py#L244
      - https://github.com/odoo/odoo/blob/9b6073c3ea6264f59fb11e6f797f4fff200c3d5d/addons/account/models/account_journal_dashboard.py#L192-L194
      - https://github.com/odoo/odoo/blob/9b6073c3ea6264f59fb11e6f797f4fff200c3d5d/addons/hr_expense/models/account_journal_dashboard.py#L30

Because of each invoice amount is converted using currency rate.

You can see the following [odoo-profiler module](https://github.com/OCA/server-tools/pull/1369) result:
 - postgresql 1

 - get_rates 1

# Desired behavior after PR is merged:

Reduce from 13.5k time executing rate query to 43 to 0, the most time consuming query is another one:
 - sql2

Reduce from python profiling cumulative time from 6.60 to 0.70 to get rates:
 - py2

Ticket 1920188

### CODE REVIEW

**Comment by nhomar** [comment]
cc @qdp-odoo

File: odoo/addons/base/models/res_currency.py [hunk 1]
```diff
@@ -274,3 +275,22 @@ def _name_search(self, name, args=None, operator='ilike', limit=100, name_get_ui
                 name = ''
                 operator = 'ilike'
         return super(CurrencyRate, self)._name_search(name, args=args, operator=operator, limit=limit, name_get_uid=name_get_uid)
+
+    @api.model
+    def create(self, vals)
```
Comment (luisg123v) [inline odoo/addons/base/models/res_currency.py]: @moylop260  
Isn't this missing a ":"?

**Comment by luisg123v** [comment]
opw-1920188

CC  @nhomar  @yaniaular

**Comment by moylop260** [comment]
Hi @mart-e 
Thanks for reply.

Yes, you are right.
The ideal way for the dasboard is to use `invoice.amount_total_company_signed ` field:
 - https://github.com/odoo/odoo/blob/62221bc3f95eba63681504283dff2d57b07ed290/addons/account/models/account_invoice.py#L319-L321

In fact, this field is not re-computed each time if you have changes in the currency rate after validated an invoice then you will have the real value with the invoice validated.

But I'm not sure if this change break the stable policies since that the result can change in the dashboard. (for master is a good proposal).

However, using ORM cache (Odoo's LRU Cache) we have an improving performance *9x* (6.60 to 0.70 to get rates)  opening dashboard without break the stable policies.

In fact, with ORM cache the validation of a invoice will be faster too since that is used the convertion rate method more than one time.

What do you think?

**Comment by nim-odoo** [comment]
Hum, actually the dashboard displays the data in the currency of the journal. So in a weird situation where the company, the journal and the invoice have a different currency, you still need this conversion. Anyway, that could probably be optimized in a way or another.

We'll check your PR, though.

**Comment by mart-e** [comment]
I like the idea to cache the method but I want to be sure it won't have a bad impact on the memory (e.g. storing in cache the conversion for the year invoices in a dashboard).
@rco-odoo what do you think?

**Comment by luisg123v** [comment]
@moylop260  
Could you check https://github.com/odoo/odoo/commit/c8ac1b40f258a2de72437ada292ec7256db36c3f? Apparently, this issue was solved there.

**Comment by moylop260** [comment]
Superseeded by https://github.com/odoo/odoo/pull/30552

**Pull request closed without merging**